src/details/ArborX_Ray.hpp

```diff
@@ -198,6 +199,216 @@ bool intersects(Ray const &ray, Box const &box)
+}
+
+// The algorithm is described in
+// Watertight Ray/Triangle Intersection
+// [1] Woop, S. et al. (2013),
+// Journal of Computer Graphics Techniques Vol. 2(1)
+// The major difference is that here we return the intersection points
+// when the ray and the triangle is coplanar.
+// In the paper, they just need the boolean return.
+KOKKOS_INLINE_FUNCTION
+bool intersection(Ray const &ray, Triangle const &triangle, float &tmin,
+                  float &tmax)
+{
+  auto dir = ray.direction();
+  // normalizing the direction vector by its largest component.
+  // if precalculated, kx, ky, kz need to be stored.
+  int kz = 0;
+  for (int i = 1; i < 3; i++)
+  {
+    float compmax = std::fabs(dir[i - 1]);
+    if (std::fabs(dir[i]) > compmax)
+    {
+      compmax = dir[i];
+      kz = i;
+    }
+  }
+  int kx = kz + 1;
+  if (kx == 3)
+    kx = 0;
+  int ky = kx + 1;
+  if (ky == 3)
+    ky = 0;
+
+  if (dir[kz] < 0.0f)
+    KokkosExt::swap(kx, ky);
+
+  Vector s;
+
+  s[2] = 1.0f / dir[kz];
+  s[0] = dir[kx] * s[2];
+  s[1] = dir[ky] * s[2];
+
+  // calculate vertices relative to ray origin
+  Vector const oA = makeVector(ray.origin(), triangle.a);
+  Vector const oB = makeVector(ray.origin(), triangle.b);
+  Vector const oC = makeVector(ray.origin(), triangle.c);
+
+  Point A;
+  Point B;
+  Point C;
+
+  // perform shear and scale of vertices
+  A[0] = oA[kx] - s[0] * oA[kz];
+  A[1] = oA[ky] - s[1] * oA[kz];
+  B[0] = oB[kx] - s[0] * oB[kz];
+  B[1] = oB[ky] - s[1] * oB[kz];
+  C[0] = oC[kx] - s[0] * oC[kz];
+  C[1] = oC[ky] - s[1] * oC[kz];
+
+  // calculate scaled barycentric coordinates
+  float u = C[0] * B[1] - C[1] * B[0];
+  float v = A[0] * C[1] - A[1] * C[0];
+  float w = B[0] * A[1] - B[1] * A[0];
+
+  // fallback to edge test using double precision
+  if (u == 0.0f || v == 0.0f || w == 0.0f)
+  {
+    double CxBy = (double)C[0] * (double)B[1];
+    double CyBx = (double)C[1] * (double)B[0];
+    u = (float)(CxBy - CyBx);
+
+    double AxCy = (double)A[0] * (double)C[1];
+    double AyCx = (double)A[1] * (double)C[0];
+    v = (float)(AxCy - AyCx);
+
+    double BxAy = (double)B[0] * (double)A[1];
+    double ByAx = (double)B[1] * (double)A[0];
+    w = (float)(BxAy - ByAx);
+  }
+
+  constexpr auto inf = KokkosExt::ArithmeticTraits::infinity<float>::value;
+  tmin = inf;
+  tmax = -inf;
+
+  // depending on the facing of the triangle
+  if ((u < 0.0f || v < 0.0f || w < 0.0f) && (u > 0.0f || v > 0.0f || w > 0.0f))
+    return false;
+
+  // calculate determinant
+  float det = u + v + w;
+
+  A[2] = s[2] * oA[kz];
+  B[2] = s[2] * oB[kz];
+  C[2] = s[2] * oC[kz];
+
+  if (det != 0.0f)
+  {
+    float t = (u * A[2] + v * B[2] + w * C[2]) / det;
+    tmax = t;
+    tmin = t;
+    return tmax >= tmin;
+  }
+  // The ray is co-planar to the triangle.
+  // Check the intersection with each edgel
+  // the rotate2D function is to make sure the ray-edge
+  // intersection check is at the plane where ray and edges
+  // are at.
```

Comment:
Still not sure if this is right. There is nothing like this in the paper, and they claim that they still detect ray-triangle edge hit. Have you tried running without?

My thinking is that if `det` is 0, that means that `(u,v,w) = (0,0,0)` (because we already tested them for the same-signess), which are not valid barycentric coordinates. So, I think a ray does not intersect with the triangle at all. So, I think
```c++
  if (det == 0)
    return false;

  float t = (u * A[2] + v * B[2] + w * C[2]) / det
  tmax = t;
  tmin = t;
  return tmax >= tmin;
```
may work.

Reply:
Hmm, does not actually work. And in the paper they actually state:
```
If this determinant is zero, the ray is co-planar to the triangle and we assume a miss.
```
I guess their definition of "hitting exactly on the edge" means something else; likely, only for non-planar intersections.

Reply:
That's the exact same problem in the other algorithm. The algorithm detects that you hit an edge except if the ray is co-planar.